File: ProjectEuler/051-100/S83.cs
using System;
using System.Collections.Generic;

namespace ProjectEuler
{
    public class S83 : ISolution
    {
        private static readonly int dim = 80;

        public string GetAnswer()
        {
            long[,] matrix = S81.GetMatrix();
            long[,] sum = new long[dim, dim];
            for (int i = 0; i < dim; i++)
            {
                for (int j = 0; j < dim; j++)
                {
                    sum[i, j] = int.MaxValue;
                }
            }

            sum[0, 0] = matrix[0, 0];
            var pointsToBeUpdated = new Queue<Tuple<int, int>>();
            pointsToBeUpdated.Enqueue(new Tuple<int, int>(0, 0));
            while (pointsToBeUpdated.Count != 0)
            {
                var point = pointsToBeUpdated.Dequeue();
                if (CanMoveToLeft(point))
                {
                    long newValue = sum[point.Item1, point.Item2] + matrix[point.Item1 - 1, point.Item2];
                    if (newValue < sum[point.Item1 - 1, point.Item2])
                    {
                        sum[point.Item1 - 1, point.Item2] = newValue;
                        pointsToBeUpdated.Enqueue(new Tuple<int, int>(point.Item1 - 1, point.Item2));
                    }
                }

                if (CanMoveToRight(point))
                {
                    long newValue = sum[point.Item1, point.Item2] + matrix[point.Item1 + 1, point.Item2];
                    if (newValue < sum[point.Item1 + 1, point.Item2])
                    {
                        sum[point.Item1 + 1, point.Item2] = newValue;
                        pointsToBeUpdated.Enqueue(new Tuple<int, int>(point.Item1 + 1, point.Item2));
                    }
                }

                if (CanUp(point))
                {
                    long newValue = sum[point.Item1, point.Item2] + matrix[point.Item1, point.Item2 - 1];
                    if (newValue < sum[point.Item1, point.Item2 - 1])
                    {
                        sum[point.Item1, point.Item2 - 1] = newValue;
                        pointsToBeUpdated.Enqueue(new Tuple<int, int>(point.Item1, point.Item2 - 1));
                    }
                }

                if (CanDown(point))
                {
                    long newValue = sum[point.Item1, point.Item2] + matrix[point.Item1, point.Item2 + 1];
                    if (newValue < sum[point.Item1, point.Item2 + 1])
                    {
                        sum[point.Item1, point.Item2 + 1] = newValue;
                        pointsToBeUpdated.Enqueue(new Tuple<int, int>(point.Item1, point.Item2 + 1));
                    }
                }
            }

            return sum[dim - 1, dim - 1].ToString();
        }

        private static bool CanMoveToLeft(Tuple<int, int> point)
        {
            return point.Item1 > 0;
        }

        private static bool CanMoveToRight(Tuple<int, int> point)
        {
            return point.Item1 < dim - 1;
        }

        private static bool CanUp(Tuple<int, int> point)
        {
            return point.Item2 > 0;
        }

        private static bool CanDown(Tuple<int, int> point)
        {
            return point.Item2 < dim - 1;
        }
    }
}

File: src/051-100/S83.cc
#include "S83.h"
#include "Matrix81.h"
#include <climits>
#include <iostream>
#include <queue>
#include <string>
#include <tuple>
#include <vector>

using namespace std;

namespace ProjectEuler
{
std::string S83::GetAnswer()
{
	const auto &matrix = MATRIX_81;
	int rows = matrix.size();
	if (rows == 0)
	{
		return "0";
	}
	int cols = matrix[0].size();

	// Dijkstra
	// dist[i][j]
	std::vector<std::vector<long long>> dist(rows, std::vector<long long>(cols, LLONG_MAX));

	// Priority queue stores {cost, r, c}
	// Min-heap
	using State = std::tuple<long long, int, int>;
	std::priority_queue<State, std::vector<State>, std::greater<State>> pq;

	dist[0][0] = matrix[0][0];
	pq.push({ dist[0][0], 0, 0 });

	int dr[] = { -1, 1, 0, 0 };
	int dc[] = { 0, 0, -1, 1 };

	while (!pq.empty())
	{
		auto [d, r, c] = pq.top();
		pq.pop();

		if (d > dist[r][c])
		{
			continue;
		}
		if (r == rows - 1 && c == cols - 1)
		{
			return std::to_string(d);
		}

		for (int i = 0; i < 4; ++i)
		{
			int nr = r + dr[i];
			int nc = c + dc[i];

			if (nr >= 0 && nr < rows && nc >= 0 && nc < cols)
			{
				if (dist[r][c] + matrix[nr][nc] < dist[nr][nc])
				{
					dist[nr][nc] = dist[r][c] + matrix[nr][nc];
					pq.push({ dist[nr][nc], nr, nc });
				}
			}
		}
	}

	return "0";
}

bool S83::CanMoveToLeft(int point) { return false; } // Unused

bool S83::CanMoveToRight(int point) { return false; } // Unused

bool S83::CanUp(int point) { return false; } // Unused

bool S83::CanDown(int point) { return false; } // Unused

} // namespace ProjectEuler
